### PR TITLE
Fix unchecked int32 indexing in some GPU kernels

### PR DESCRIPTION
**Pull request by benbarsdell**
- Many op kernels use the `To32Bit()` function, which does not do any bounds checking and can cause silent corruption due to int32 overflow.
  E.g., `tf.zeros/ones/fill` silently give incorrect results when the number of elements is >= 2**31.
- This commit adds a new helper function `MaybeWith32BitIndexing()` that checks the indexing bounds and converts tensors to 32-bit only if it is safe to do so (and only for GPU devices). This fixes the cases that were previously broken, and simplifies the ones that already had explicit bounds checks.

cc @nluehr @sanjoy

### CODE REVIEW

**Review by sanjoy — CHANGES_REQUESTED** [review]
We have an ongoing engagement with @kushanam and @sherhut to do this without bloating the binary size by JIT compiling the rarely used codepaths.  So I would rather invest into that effort unless we've already judged the JIT-based approach to be infeasible.

File: tensorflow/core/framework/tensor_types.h [hunk 1]
```diff
@@ -119,7 +155,44 @@ typename TTypes<typename TensorType::Scalar,
...
+struct MaybeWith32BitIndexingImpl<Eigen::GpuDevice> {
+  template <typename Func, typename... Args>
+  void operator()(Func func, Args&&... args) const {
+    auto all = [](const auto&... bool_vals) {
+      bool result = true;
+      int unpack[] = {0, (result = result && bool_vals, 0)...};
+      (void)unpack;
+      return result;
+    };
+    if (all((SafeFor32BitIndexing(std::forward<Args>(args)))...)) {
```
[the remark this answers is not in the document]
Comment (benbarsdell) [reply tensorflow/core/framework/tensor_types.h]: Nope, fixed.

File: tensorflow/core/framework/tensor_types.h [hunk 2]
```diff
@@ -119,7 +155,44 @@ typename TTypes<typename TensorType::Scalar,
...
+  }
+};
+
+template <>
+struct MaybeWith32BitIndexingImpl<Eigen::GpuDevice> {
+  template <typename Func, typename... Args>
+  void operator()(Func func, Args&&... args) const {
+    auto all = [](const auto&... bool_vals) {
+      bool result = true;
+      int unpack[] = {0, (result = result && bool_vals, 0)...};
```
[the remark this answers is not in the document]
Comment (benbarsdell) [reply tensorflow/core/framework/tensor_types.h]: Good idea, done.

**Comment by benbarsdell** [comment]
> We have an ongoing engagement with @kushanam and @sherhut to do this without bloating the binary size by JIT compiling the rarely used codepaths. So I would rather invest into that effort unless we've already judged the JIT-based approach to be infeasible.

OK I was wondering about that. If the timeline is not too long then that would be the way to go.

**Comment by benbarsdell** [comment]
My take on this is that we can close it if the new JIT kernels work will be enabled before the release of 2.7.
If that won't make 2.7, then we should either merge this PR or modify it to simply throw a fatal error when 32-bit indexing is insufficient.

For reference, I checked the effect of this PR on binary size (_pywrap_tensorflow_internal.so) in my build:
Before: 547MB
After:  566MB (+3.5%)

**Comment by sherhut** [comment]
We should land this change independently of the kernel generator work, as it also covers other operations (like reductions) that are not currently covered by kernel generator.

I suspect that the (relatively) small change in size is due to us not generating all the cwise operations via the eigen code path anymore, so the generation of two kernels due to `cwise_cpu_common` does not hit us.

**Review by sanjoy — APPROVED** [review]

**Pull request merged**